packages/@glimmer/application/src/components/component-managers/template-only.ts

```diff
@@ -0,0 +1,173 @@
+  createCaller: false,
+  createInstance: true,
+  dynamicLayout: false,
+  dynamicScope: false,
+  dynamicTag: true,
+  elementHook: true,
+  prepareArgs: false,
+  updateHook: true,
+  wrapped: false,
+};
+
+export class ComponentStateBucket {
+  public name: string;
+
+  constructor(public args: CapturedArguments) {
+    this.args = args;
+  }
+}
+
+const EMPTY_SELF = new RootReference(null);
+
+/**
+ * For performance reasons, we want to avoid instantiating component buckets for
+ * components that don't have an associated component class that we would need
+ * instantiate and invoke lifecycle hooks on.
+ *
+ * In development mode, however, we need to track some state about the component
+ * in order to produce more useful error messages. This
+ * TemplateOnlyComponentDebugBucket is only created in development mode to hold
+ * that state.
+ */
+export class TemplateOnlyComponentDebugBucket {
+  constructor(public definition: TemplateOnlyComponentDefinition) {}
+}
+
+export interface ExtendedTemplateMeta {
+  specifier: string;
+  managerId?: string;
+}
+
+export interface CompilableRuntimeResolver extends RuntimeResolver<ExtendedTemplateMeta> {
+  compileTemplate(name: string, layout: Option<number>): Invocation;
+}
+
+export interface EnvironmentWithOwner extends Environment {
+  getOwner(): Owner;
+  setOwner(obj: Object, owner: Owner): void;
+}
+
+export default class TemplateOnlyComponentManager
+  implements
+    VMComponentManager<TemplateOnlyComponentDebugBucket | null, ComponentDefinitionState>,
+    WithJitStaticLayout<
+      TemplateOnlyComponentDebugBucket | null,
+      TemplateOnlyComponentDefinitionState,
+      JitRuntimeResolver
+    >,
+    WithAotStaticLayout<
+      TemplateOnlyComponentDebugBucket | null,
+      TemplateOnlyComponentDefinitionState,
+      AotRuntimeResolver
+    > {
+  static create(): TemplateOnlyComponentManager {
+    return new TemplateOnlyComponentManager();
+  }
+
+  getCapabilities(): ComponentCapabilities {
+    return CAPABILITIES;
+  }
+
+  getJitStaticLayout({ definition }: TemplateOnlyComponentDefinitionState): CompilableProgram {
+    return definition.template.asLayout();
+  }
+
+  getAotStaticLayout({ definition }: TemplateOnlyComponentDefinitionState): Invocation {
+    const { handle, symbolTable } = definition;
+    return {
+      handle,
+      symbolTable
+    };
+  }
+
+  create(
+    _env: Environment,
+    state: TemplateOnlyComponentDefinitionState
+  ): TemplateOnlyComponentDebugBucket | void {
+    // In development mode, save off state needed for error messages. This will
+    // get stripped in production mode and no bucket will be instantiated.
+    return DEBUG ? new TemplateOnlyComponentDebugBucket(state.definition) : null;
+  }
+
+  getSelf(bucket: TemplateOnlyComponentDebugBucket): PathReference {
+    return DEBUG ? new TemplateOnlyComponentDebugReference(bucket.definition.state.name) : EMPTY_SELF;
+  }
+
+  getTag(): Tag {
+    return CONSTANT_TAG;
```

Comment:
I think this needs to be args.tag actually

Reply:
Are you sure? I thought the tag returned from `getTag()` was only used to determine whether the update hooks should run. Given that TOs don't have any lifecycle hooks to call, are there other cases relying on this invalidating when args change?